### PR TITLE
segment interface refactor for proposal 2965

### PR DESCRIPTION
**Pull request by javasoze**
This PR performs the refactor on the Segment interface described in proposal 2965: https://github.com/druid-io/druid/issues/2965

### CODE REVIEW

**Comment by drcrallen** [comment]
Hi @javasoze thanks for the PR, the discussion in #2987 is about getting rid of 

```java
    public QueryableIndex asQueryableIndex();
    public StorageAdapter asStorageAdapter();
```

My understanding of the state of the discussion is that having only `<T> T as(Class<T> clazz);` is fine as long as there is clear javadoc explanations on the limitations of using anything other than `StorageAdapter` or `QueryableIndex`.

As such the `Segment` interface should have the methods in question removed, and the javadoc for the functionality of `as` needs expanded greatly.


Also the import formatting seems off, can you run it against https://github.com/druid-io/druid/blob/master/druid_intellij_formatting.xml ?

**Comment by gianm** [comment]
@drcrallen I think you meant #2965 not #2987

**Comment by drcrallen** [comment]
@gianm yes fixed, silly git auto-filling

File: server/src/test/java/io/druid/server/coordination/ServerManagerTest.java [hunk 1]
```diff
@@ -82,6 +61,30 @@
 import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
 
+import org.joda.time.Interval;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import com.fasterxml.jackson.core.type.TypeReference;
```
Comment (fjy) [inline server/src/test/java/io/druid/server/coordination/ServerManagerTest.java]: are all these imports required?

Comment (fjy) [inline server/src/test/java/io/druid/server/coordination/ServerManagerTest.java]: nvm, it is the IDE reorganizing things

**Comment by fjy** [comment]
👍

File: processing/src/test/java/io/druid/query/aggregation/AggregationTestHelper.java [hunk 2]
```diff
@@ -387,12 +390,12 @@ public void createIndex(
 
   public Sequence<Row> runQueryOnSegments(final List<File> segmentDirs, final Query query)
   {
-    final List<Segment> segments = Lists.transform(
+    final List<AbstractSegment> segments = Lists.transform(
```
Comment (cheddar) [inline processing/src/test/java/io/druid/query/aggregation/AggregationTestHelper.java]: Why are these changes required?  I think that keeping it at just `Segment` should be fine (and more correct, 'cause it's totally possible that someone chooses to implement `Segment` without using `AbstractSegment`).

**Comment by cheddar** [comment]
This generally LGTM.  Charles' comment about totally removing the other two methods I think is legit.  I see two paths forward:

1. We could expand this PR to remove those methods as well and update all of the individual call sites.  That would mean that the PR cannot actually be merged until we have a backwards incompatible release, 'cause it's possible that someone has implemented their own Query flow.
2. We could accept this PR as is for now and essentially declare the `Segment` interface to be experimental and in-flux.  When we have it solidified, we do another PR to remove the other methods and declare things good.

I'm generally in favor of Option 1, if we are ok with doing a backwards incompatible release in the very near future (i.e. this gets merged without sitting here in the PR queue for a really long time).  If we are not ok with that, then I think we need to take approach 2.

**Comment by gianm** [comment]
I'm in favor of 2 for the time being, it seems more straightforward to merge that in for 0.9.2. Then once the interface changes have proven they can enable real world usage, we could do an expanded adjustment in the next major release.

**Comment by javasoze** [comment]
@drcrallen Sorry about that, I forgot to load the eclipse import order file. Now fixed.
@cheddar @gianm I am also in favor of option 2) because I am normally in favor of incremental and isolated changes, especially API related changes.

File: processing/src/test/java/io/druid/query/MultiValuedDimensionTest.java [hunk 3]
```diff
@@ -171,8 +174,8 @@ public void testGroupByNoFilter() throws Exception
 
     result = helper.runQueryOnSegmentsObjs(
         ImmutableList.of(
```
Comment (gianm) [inline processing/src/test/java/io/druid/query/MultiValuedDimensionTest.java]: [minor] `ImmutableList.<Segment>of` should also work and avoid the cast

**Comment by gianm** [comment]
Suggest this javadoc:

```java
  /**
   * Request an implementation of a particular interface.
   *
   * If the passed-in interface is {@link QueryableIndex} or {@link StorageAdapter}, then this method behaves
   * identically to {@link #asQueryableIndex()} or {@link #asStorageAdapter()}. Other interfaces are only
   * expected to be requested by callers that have specific knowledge of extra features provided by specific
   * segment types. For example, an extension might provide a custom Segment type that can offer both
   * StorageAdapter and some new interface. That extension can also offer a Query that uses that new interface.
   *
   * @param clazz desired interface
   * @param <T> desired interface
   * @return instance of clazz, or null if the interface is not supported by this segment
   */
  public <T> T as(Class<T> clazz);
```

👍  otherwise.

**Comment by drcrallen** [comment]
In addition to what @gian wrote, please add something to the effect of

```
Implementations which accept classes other than {@link QueryableIndex} or {@link StorageAdapter} are limited to using those classes within the extension. This means that one extension cannot rely on the `Segment.as` behavior of another extension.
```

**Comment by fjy** [comment]
@javasoze we can get this in the 0.9.1 release if you can update the javadoc

File: processing/src/test/java/io/druid/query/aggregation/AggregationTestHelper.java (hunk 2 above)
Comment (javasoze) [inline processing/src/test/java/io/druid/query/aggregation/AggregationTestHelper.java]: You are right, they are not, fixed.

**Comment by gianm** [comment]
👍  LGTM

**Comment by cheddar** [comment]
👍
